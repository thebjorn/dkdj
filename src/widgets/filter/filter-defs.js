import {UIWidget} from "../../widgetcore/ui-widget";
import dk from "../../dk-obj";
import {CheckboxSelectWidget, RadioSelectWidget, SelectWidget} from "../../forms/widgets";
import {dkconsole} from "../../lifecycle/dkboot/dk-console";

/**
 * A single filter definition.
 * Think of it as a filter for a single field.
 * Multiple filter definitions will normally make up the filter for a datasource.
 */
export class FilterDefBase extends UIWidget {
    constructor(...args) {
        super({
            select_multiple: false,
            name: null,
            label: "",
            values: null,
            structure: {
                classes: ['filterdef'],
                filterbox: {
                    classes: [],
                    filterheader: {
                        template: 'header',
                        filtertitle: {
                            template: 'h3',
                            text: ""
                        }
                    },
                    filtercontent: {
                        classes: ['filter', 'content']
                    }
                }
            }
        }, ...args);

        this.structure.filterbox.classes.push(this.name);
        this.structure.filterbox.filterheader.filtertitle.text = this.label || this.name;
        // this is set to true for select-one/-many, so the data source's get_foo_filter_values
        // method will be called, you must set it to true manually in any other filter, 
        // e.g. a custom filter.
        this.needs_options = false;
    }

    get value() {
        return this.input ? this.input.value : {};  // should be this.widget, but clashes with dk.UIWidget.widget()
    }

    handlers() {
        if (this.input) dk.on(this.input, 'change', (evt, widget) => {
            // eslint-disable-next-line no-console
            // console.info("DK:ON:CHANGE:", evt, widget);
            this.trigger('change', this.value);
        });
    }

    draw(opts) {
        // console.log("FILTERDEF:BASE:DRAW:", opts);
        if (this.input && this.input.draw) {
            this.input.draw(opts);
        }
    }

    fetch_options() {
        if (this.needs_options) {
            if (this.datafilter && this.datafilter.dataset) {
                this.datafilter.dataset.get_filter_data(this.name, opts => {
                    this.draw(opts);
                    // eslint-disable-next-line no-console
                    dkconsole.info('fetch_options-1', opts);
                });
            } else {
                // eslint-disable-next-line no-console
                dkconsole.info('fetch_options-2', this.datafilter.dataset);
            }
        }
    }
}

/**
 * Filters can construct their own UI, using the construct method,
 * which takes a location. When generating the UI you must:
 *
 *  - create a `this.widget` property (the filter system listens for
 *    notifications on this widget.
 *  - set `this.value` to the correct filter value (this can be a
 *    compound value, but must be _JSON_ serializable).
 *  - call `dk.publish(this.widget, 'change', this.widget)` whenever
 *    the filter value changes.
 */
export class CustomFilterDef extends FilterDefBase {
    constructor(...args) {
        const props = Object.assign({}, ...args);
        const name = props.name;
        delete props.name;
        super(props);
        this._name = name;
    }

    get value() {
        return this._value;
    }

    set value(val) {
        this._value = val;
    }

    construct() {
        this.input = this.construct_filter(this.filterbox.filtercontent, this);
    }

}

export class SelectOneFilterDef extends FilterDefBase {
    constructor(...args) {
        super({
            select_multiple: false
        }, ...args);
        this.needs_options = true;
    }

    construct() {
        this.input = RadioSelectWidget.create_on(this.filterbox.filtercontent, {
            options: this.values,
            name: this.name,
            label: this.label
        });
        if (!this.values) this.fetch_options();
    }
}

export class SelectMultipleFilterDef extends FilterDefBase {
    constructor(...args) {
        super({
            select_multiple: true
        }, ...args);
        this.needs_options = true;
    }

    construct() {
        // console.log("SELECT:MULTIPLE:FILTERDEFS:CONSTRUCT:VALUES:", this.values);
        this.input = CheckboxSelectWidget.create_on(this.filterbox.filtercontent, {
            options: this.values,
            name: this.name,
            label: this.label
        });
        if (!this.values) this.fetch_options();
    }
}


export class DropdownFilterDef extends FilterDefBase {
    constructor(...args) {
        super({
            select_multiple: false,
            structure: {
                classes: ['filterdef'],
                filterbox: {
                    classes: [],
                    filterheader: {
                        template: 'header',
                        filtertitle: {
                            template: 'h3',
                            text: ""
                        }
                    },
                    filtercontent: {
                        classes: ['filter', 'content']
                    }
                }
            },
            template: {
                filtercontent: 'div',
            }
        }, ...args);
        this.needs_options = true;
    }

    construct() {
        this.input = SelectWidget.create_inside(this.filterbox.filtercontent, {
            options: this.values,
            name: this.name,
            label: this.label
        });
        if (!this.values) this.fetch_options();
    }
}
