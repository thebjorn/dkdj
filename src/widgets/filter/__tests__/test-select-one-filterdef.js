import $ from 'jquery';
import page from "../../../widgetcore/dk-page";
import {SelectOneFilterDef} from "../filter-defs";


test("test-select-one-filterdef", () => {
    document.body.innerHTML = `
        <div id="work"></div>
    `;
    const work = $('#work');
    page.initialize(document);
    
    const sofd = SelectOneFilterDef.create_inside(work, {
        name: 'myname',
        label: 'My Label',
        values: [1, 2, 3]
    });
    
    console.log(work.html());
    expect(work.html()).toMatchSnapshot();
    
    expect(sofd.value).toEqual([]);
    
    work.find(':radio:eq(0)').click();
    expect(sofd.value).toEqual(["1"]);

    work.find(':radio:eq(1)').click();
    expect(sofd.value).toEqual(["2"]);

    // expect(utidy(work.html())).toEqual(utidy(`
    //
    // `));
});
