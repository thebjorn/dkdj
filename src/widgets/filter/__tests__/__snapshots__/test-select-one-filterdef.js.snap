// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test-select-one-filterdef 1`] = `
"<div id=\\"select-one-filter-def-1\\" class=\\"SelectOneFilterDef filterdef\\"><div id=\\"dk-bx-1\\" class=\\"myname filterbox\\"><header id=\\"dk-bx-2\\" class=\\"filterheader\\"><h3 id=\\"dk-bx-3\\" class=\\"filtertitle\\">My Label</h3></header><div id=\\"dk-bx-4\\" class=\\"filter content filtercontent RadioSelectWidget\\"><label id=\\"myname-1-label\\" for=\\"myname-1\\"><input type=\\"radio\\" name=\\"myname\\" id=\\"myname-1\\" value=\\"1\\">1</label>
        <label id=\\"myname-2-label\\" for=\\"myname-2\\"><input type=\\"radio\\" name=\\"myname\\" id=\\"myname-2\\" value=\\"2\\">2</label>
        <label id=\\"myname-3-label\\" for=\\"myname-3\\"><input type=\\"radio\\" name=\\"myname\\" id=\\"myname-3\\" value=\\"3\\">3</label>
        </div></div></div>"
`;
